File: 0008.StringtoInteger(atoi).py
class Solution(object):
    def myAtoi(self, str):
        """
        :type str: str
        :rtype: int
        """
        if not str:
            return 0
        ans = 0
        s = str.strip()
        sign = 1
        if s[0] == '+':
            s = s[1:]
        elif s[0] == '-':
            sign = -1
            s = s[1:]
        for ch in s:
            n = ord(ch) - ord('0')
            if n >= 0 and n <= 9:
                ans = 10*ans + n
            else:
                break
        ans *= sign
        ans = min(ans, 2**31-1)
        ans = max(ans, -2**31)
        return ans

File: 0008.StringtoInteger(atoi).go
func myAtoi(str string) int {
    n := len(str)
    i := 0
    for ; i < n && str[i] == ' '; i++ {}
    if i >= n {
        return 0
    }
    sign := 1
    if str[i] == '-' {
        sign = -1
        i++
    } else if str[i] == '+' {
        i++
    }
    num := 0
    for ; i < n && str[i] >= '0' && str[i] <= '9'; i++ {
        d := int(str[i]-'0')
        if num > math.MaxInt32/10 || (num == math.MaxInt32/10 && d > math.MaxInt32%10) {
            if sign == 1 {
                return math.MaxInt32
            }
            return math.MinInt32
        }
        num = 10*num + d
    }
    return sign*num
}